xapian-core/weight/tfidfweight.cc

```diff
@@ -76,12 +78,18 @@ TfIdfWeight::TfIdfWeight(const std::string &normals, double slope, double delta)
     need_stat(WDF);
     need_stat(WDF_MAX);
     need_stat(WQF);
-    need_stat(DOC_LENGTH);
-    need_stat(DOC_LENGTH_MIN);
-    need_stat(DOC_LENGTH_MAX);
-    need_stat(UNIQUE_TERMS);
     if (normalizations[0] == 'P' || normalizations[1] == 'P') {
 	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(DOC_LENGTH_MIN);
+    }
+    else if (normalizations[0] == 'L') {
+	need_stat(DOC_LENGTH);
+	need_stat(DOC_LENGTH_MIN);
+	need_stat(DOC_LENGTH_MAX);
+	need_stat(UNIQUE_TERMS);
+	if (normalizations[1] == 'P')
+	    need_stat(AVERAGE_LENGTH);
```

Comment:
I would just make this a separate `if (normalizations[0] == 'L') {` and drop the duplicated handling for `normalizations[1] == 'P'` from it.  That way the knowledge about `(normalizations[1] == 'P')` isn't duplicated (https://en.wikipedia.org/wiki/Don't_repeat_yourself).

The implementation of `need_stat(flag);` is carefully written to just inline `stats_needed = stat_flags(stats_needed | flag);` so a contiguous block of such calls with constant values should optimise down to a single bitwise or with a constant.  So this code is really:

```
if (normalizations[0] == 'P' || normalizations[1] == 'P') {
    stats_needed |= SOME_CONSTANT1;
} else if (normalizations[0] == 'L') {
    stats_needed |= SOME_CONSTANT2;
    if (normalizations[1] == 'P')
        stats_needed |= SOME_CONSTANT3;
}
```

And my suggested version would be:

```
if (normalizations[0] == 'P' || normalizations[1] == 'P') {
    stats_needed |= SOME_CONSTANT1;
}
if (normalizations[0] == 'L') {
    stats_needed |= SOME_CONSTANT2;
}
```